torchvision/prototype/datasets/_builtin/cub.py

```diff
@@ -0,0 +1,233 @@
+                sha256="aeacbd5e3539ae84ea726e8a266a9a119c18f055cd80f3836d5eb4500b005428",
+                decompress=True,
+            )
+            images = HttpResource(
+                "http://www.vision.caltech.edu/visipedia-data/CUB-200/images.tgz",
+                sha256="2a6d2246bbb9778ca03aa94e2e683ccb4f8821a36b7f235c0822e659d60a803e",
+                decompress=True,
+            )
+            anns = HttpResource(
+                "http://www.vision.caltech.edu/visipedia-data/CUB-200/annotations.tgz",
+                sha256="c17b7841c21a66aa44ba8fe92369cc95dfc998946081828b1d7b8a4b716805c1",
+                decompress=True,
+            )
+            return [split, images, anns]
+
+    def _2011_classify_archive(self, data: Tuple[str, Any]) -> Optional[int]:
+        path = pathlib.Path(data[0])
+        if path.parents[1].name == "images":
+            return 0
+        elif path.name == "train_test_split.txt":
+            return 1
+        elif path.name == "images.txt":
+            return 2
+        elif path.name == "bounding_boxes.txt":
+            return 3
+        else:
+            return None
+
+    _2011_SPLIT_MAP = {
+        "0": "test",
+        "1": "train",
+    }
+
+    def _2011_filter_split(self, row: List[str], *, split: str) -> bool:
+        _, split_id = row
+        return self._2011_SPLIT_MAP[split_id] == split
+
+    def _2011_segmentation_key(self, data: Tuple[str, Any]) -> str:
+        path = pathlib.Path(data[0])
+        return path.with_suffix(".jpg").name
+
+    def _2011_decode_ann(
+        self,
+        data: Tuple[str, Tuple[List[str], Tuple[str, io.IOBase]]],
+        *,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> Dict[str, Any]:
+        _, (bounding_box_data, segmentation_data) = data
+        segmentation_path, segmentation_buffer = segmentation_data
+        return dict(
+            bounding_box=BoundingBox([float(part) for part in bounding_box_data[1:]], format="xywh"),
+            segmentation_path=segmentation_path,
+            segmentation=Feature(decoder(segmentation_buffer)) if decoder else segmentation_buffer,
+        )
+
+    def _2010_split_key(self, data: str) -> str:
+        return data.rsplit("/", maxsplit=1)[1]
+
+    def _2010_anns_key(self, data: Tuple[str, io.IOBase]) -> Tuple[str, Tuple[str, io.IOBase]]:
+        path = pathlib.Path(data[0])
+        return path.with_suffix(".jpg").name, data
+
+    def _2010_decode_ann(
+        self, data: Tuple[str, Tuple[str, io.IOBase]], *, decoder: Optional[Callable[[io.IOBase], torch.Tensor]]
+    ) -> Dict[str, Any]:
+        _, (path, buffer) = data
+        content = read_mat(buffer)
+        return dict(
+            ann_path=path,
+            bounding_box=BoundingBox(
+                [int(content["bbox"][coord]) for coord in ("left", "bottom", "right", "top")], format="xyxy"
+            ),
+            segmentation=Feature(content["seg"]),
```

Comment:
Could you help me understand why we use the `decoder` for the `Feature(...)` in `_2011_decode_ann`, but not in this function?

Reply:
The `decoder` is used to turn an open file handle into a the pixel values as tensor. In this case the raw pixel values are already present as a `numpy.ndarray` so there is nothing left to decode.